### ISSUE TITLE
Autodecorator runs on foreign objects

### ISSUE THREAD

**Issue by herczy** [balabit/typesafety#13]
When importing an object from an external, unchecked module into module `A` and importing that in module `B` the object will be autodecorated. Consider the following code:

Module A

``` python
from mock import Mock
```

Module B

``` python
from A import Mock
```

When the tests are run for module B, we get the following error message:

```
  File "/usr/lib/python3.2/inspect.py", line 834, in getfullargspec
    raise TypeError('{!r} is not a Python function'.format(func))
TypeError: 'The value to be returned when the mock is called.' is not a Python function
```

**Comment by attilammagyar** [comment]
If I understand correctly, this issue might be a bit tricky.

On one hand, if the import in `module B` is accidental, then this behaviour may be unexpected.

On the other hand, if `module A` provides the entity in question as part of its public interface, then if you are intentionally type-checking `module A`, then you probably want it type-checked in its entirety, regardless of it defining components of its API on its own or importing it from elsewhere. Probably a proper solution would be to provide a way to skip type-checks for a given set of identifiers, similar to `__all__`:

``` python
from other_module import ClassUsedByThisModuleButNotPartOfItsApi


class ClassProvidedByThisModule:
    pass


TYPESAFETY_IGNORE = ["ClassUsedByThisModuleButNotPartOfItsApi"]
```

or maybe:

``` python
TYPESAFETY_IGNORE = [ClassUsedByThisModuleButNotPartOfItsApi]
```

**Comment by herczy** [comment]
That is not really the problem. The problem is, that when we typecheck module A, we know that `Mock` does not need to be checked, since the activation filter tells the autodecorator not to. However when module B is checked, typesafety will try to decorate Mock too, despite the fact that it comes from somewhere else.

The very fact that the filter function given to `activate` tells us that it does not need to be checked. I think we should simply make the `Autodecorator` class track the origins of an object in a module. I think it can be done easily, will post a PoC.

**Comment by herczy** [comment]
Added a feature file that describes the problem a little more exactly as we have here.

The gist of it is that I don't think that objects imported from external modules (that is, modules that are not in `typesafety.activate`'s filter condition) should not be checked. If you agree, I will implement the fix too.

**Comment by herczy** [comment]
Meanwhile I figured out what is the root of the problem: when we are running typesafety, we autodecorate anything that we've loaded (see `ModuleFinder`). So a "foreign" module will be decorated if it needs to be when the finder accepts it. So we only need to decorate "native" module objects. See my PR for details.

There is also a problem associated with this: if we only want to decorate the package `__init__.py` file, we cannot do it, since we explicitly decorated imported submodules.
